src/Elastic.Transport/Responses/DefaultResponseFactory.cs

```diff
@@ -0,0 +1,152 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Net.NetworkInformation;
+using System.Threading;
+using System.Threading.Tasks;
+using Elastic.Transport.Diagnostics;
+using Elastic.Transport.Extensions;
+
+namespace Elastic.Transport;
+
+/// <summary>
+/// A <see cref="ResponseFactory"/> which resolves an <see cref="IResponseBuilder"/> for each response being created.
+/// </summary>
+/// <remarks>
+/// Create an instance of the factory using the provided configuration.
+/// </remarks>
+internal sealed class DefaultResponseFactory(ITransportConfiguration transportConfiguration) : ResponseFactory
+{
+	private readonly ConcurrentDictionary<Type, IResponseBuilder> _resolvedBuilders = new()
```

Comment:
The builder instances in the dictionary could be `static` I think - they don't seem to maintain a state.